File: rust/src/parser.rs
use crate::ast::{
    EachBlockNode, IfBlockNode, IncludeNode, Node, Template, TextNode, UnlessBlockNode,
    UnsecureBlockNode, VariableNode,
};
use crate::error::{Location, NatsuzoraError, Result};
use crate::token::{Token, TokenKind};
use crate::validator;

/// Recursive descent parser for Natsuzora templates
pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    /// Create a new parser from a token stream
    pub fn new(tokens: Vec<Token>) -> Self {
        Self { tokens, pos: 0 }
    }

    /// Parse the token stream into an AST
    pub fn parse(&mut self) -> Result<Template> {
        let nodes = self.parse_nodes()?;
        Ok(Template {
            nodes,
            location: Location::new(1, 1),
        })
    }

    fn parse_nodes(&mut self) -> Result<Vec<Node>> {
        let mut nodes = Vec::new();
        while !self.is_eof() && !self.block_close(None)? {
            nodes.push(self.parse_node()?);
        }
        Ok(nodes)
    }

    fn parse_node(&mut self) -> Result<Node> {
        match &self.current_token().kind {
            TokenKind::Text(_) => self.parse_text(),
            TokenKind::Open => self.parse_mustache(),
            _ => self.unexpected_token(None),
        }
    }

    fn parse_text(&mut self) -> Result<Node> {
        let token = self.consume_text()?;
        match token.kind {
            TokenKind::Text(content) => Ok(Node::Text(TextNode {
                content,
                location: token.location,
            })),
            _ => Err(NatsuzoraError::ParseError {
                message: format!("Internal error: expected text token, got {:?}", token.kind),
                location: token.location,
            }),
        }
    }

    fn parse_mustache(&mut self) -> Result<Node> {
        let open_token = self.consume(TokenKind::Open)?;

        // Check for illegal whitespace before special characters (#, /, >)
        if matches!(self.current_token().kind, TokenKind::Whitespace(_)) {
            let saved_pos = self.pos;
            self.skip_whitespace();
            let is_special = matches!(
                self.current_token().kind,
                TokenKind::Hash | TokenKind::Slash | TokenKind::Gt
            );
            if is_special {
                return Err(NatsuzoraError::ParseError {
                    message: format!(
                        "Whitespace not allowed after '{{[' before '{:?}'",
                        self.current_token().kind
                    ),
                    location: open_token.location,
                });
            }
            self.pos = saved_pos;
        }

        self.skip_whitespace();

        match &self.current_token().kind {
            TokenKind::Hash => self.parse_block_open(),
            TokenKind::Slash => self.unexpected_token(Some("Unexpected block close")),
            TokenKind::Gt => self.parse_include(),
            _ => self.parse_variable_node(),
        }
    }

    fn parse_block_open(&mut self) -> Result<Node> {
        self.consume(TokenKind::Hash)?;
        self.skip_whitespace();

        match &self.current_token().kind {
            TokenKind::KwIf => self.parse_if_block(),
            TokenKind::KwUnless => self.parse_unless_block(),
            TokenKind::KwEach => self.parse_each_block(),
            TokenKind::KwUnsecure => self.parse_unsecure_block(),
            TokenKind::KwElse => self.unexpected_token(Some("Unexpected 'else' without 'if'")),
            _ => self.unexpected_token(None),
        }
    }

    fn parse_if_block(&mut self) -> Result<Node> {
        let token = self.consume(TokenKind::KwIf)?;
        let location = token.location;

        self.consume_required_whitespace()?;
        let condition = self.parse_path()?;
        self.skip_whitespace();
        self.consume(TokenKind::Close)?;

        let then_nodes = self.parse_if_body()?;
        let else_nodes = if self.else_open()? {
            self.consume_else()?;
            Some(self.parse_if_body()?)
        } else {
            None
        };

        self.consume_block_close(Some(TokenKind::KwIf))?;

        Ok(Node::IfBlock(IfBlockNode {
            condition,
            then_nodes,
            else_nodes,
            location,
        }))
    }

    fn parse_if_body(&mut self) -> Result<Vec<Node>> {
        let mut nodes = Vec::new();
        while !self.block_close(Some(TokenKind::KwIf))? && !self.else_open()? {
            nodes.push(self.parse_node()?);
        }
        Ok(nodes)
    }

    fn parse_unless_block(&mut self) -> Result<Node> {
        let token = self.consume(TokenKind::KwUnless)?;
        let location = token.location;

        self.consume_required_whitespace()?;
        let condition = self.parse_path()?;
        self.skip_whitespace();
        self.consume(TokenKind::Close)?;

        let body_nodes = self.parse_unless_body()?;

        self.consume_block_close(Some(TokenKind::KwUnless))?;

        Ok(Node::UnlessBlock(UnlessBlockNode {
            condition,
            body_nodes,
            location,
        }))
    }

    fn parse_unless_body(&mut self) -> Result<Vec<Node>> {
        let mut nodes = Vec::new();
        while !self.block_close(Some(TokenKind::KwUnless))? {
            nodes.push(self.parse_node()?);
        }
        Ok(nodes)
    }

    fn else_open(&mut self) -> Result<bool> {
        if !matches!(self.current_token().kind, TokenKind::Open) {
            return Ok(false);
        }

        let saved_pos = self.pos;
        self.advance_token(); // open

        // If there's whitespace after {[, it cannot be {[#else]}
        if matches!(self.current_token().kind, TokenKind::Whitespace(_)) {
            self.pos = saved_pos;
            return Ok(false);
        }

        let result = if matches!(self.current_token().kind, TokenKind::Hash) {
            self.advance_token(); // hash
            self.skip_whitespace();
            matches!(self.current_token().kind, TokenKind::KwElse)
        } else {
            false
        };

        self.pos = saved_pos;
        Ok(result)
    }

    fn consume_else(&mut self) -> Result<()> {
        self.consume(TokenKind::Open)?;
        self.skip_whitespace();
        self.consume(TokenKind::Hash)?;
        self.skip_whitespace();
        self.consume(TokenKind::KwElse)?;
        self.skip_whitespace();
        self.consume(TokenKind::Close)?;
        Ok(())
    }

    fn parse_each_block(&mut self) -> Result<Node> {
        let token = self.consume(TokenKind::KwEach)?;
        let location = token.location;

        self.consume_required_whitespace()?;
        let collection = self.parse_path()?;
        self.consume_required_whitespace()?;
        self.consume(TokenKind::KwAs)?;
        self.consume_required_whitespace()?;
        let item_name = self.parse_identifier_with_validation()?;

        self.skip_whitespace();
        let index_name = if matches!(self.current_token().kind, TokenKind::Comma) {
            self.consume(TokenKind::Comma)?;
            self.skip_whitespace();
            let idx = self.parse_identifier_with_validation()?;

            if item_name == idx {
                return Err(NatsuzoraError::ParseError {
                    message: format!("Item and index cannot have the same name: '{}'", item_name),
                    location,
                });
            }

            Some(idx)
        } else {
            None
        };

        self.skip_whitespace();
        self.consume(TokenKind::Close)?;

        let body_nodes = self.parse_each_body()?;

        self.consume_block_close(Some(TokenKind::KwEach))?;

        Ok(Node::EachBlock(EachBlockNode {
            collection,
            item_name,
            index_name,
            body_nodes,
            location,
        }))
    }

    fn parse_each_body(&mut self) -> Result<Vec<Node>> {
        let mut nodes = Vec::new();
        while !self.block_close(Some(TokenKind::KwEach))? {
            nodes.push(self.parse_node()?);
        }
        Ok(nodes)
    }

    fn parse_unsecure_block(&mut self) -> Result<Node> {
        let token = self.consume(TokenKind::KwUnsecure)?;
        let location = token.location;

        self.skip_whitespace();
        self.consume(TokenKind::Close)?;

        let mut body_nodes = Vec::new();
        while !self.block_close(Some(TokenKind::KwUnsecure))? {
            body_nodes.push(self.parse_node()?);
        }

        self.consume_block_close(Some(TokenKind::KwUnsecure))?;

        Ok(Node::UnsecureBlock(UnsecureBlockNode {
            nodes: body_nodes,
            location,
        }))
    }

    fn parse_include(&mut self) -> Result<Node> {
        let token = self.consume(TokenKind::Gt)?;
        let location = token.location;

        self.skip_whitespace();
        let name = self.parse_include_name()?;
        let args = self.parse_include_args()?;
        self.skip_whitespace();
        self.consume(TokenKind::Close)?;

        Ok(Node::Include(IncludeNode {
            name,
            args,
            location,
        }))
    }

    fn parse_include_name(&mut self) -> Result<String> {
        let token = self.current_token().clone();
        let name = match &token.kind {
            TokenKind::Ident(s) if s.starts_with('/') => s.clone(),
            _ => {
                return Err(NatsuzoraError::ParseError {
                    message: "Include name must start with '/'".to_string(),
                    location: token.location,
                });
            }
        };

        validator::validate_include_name_syntax(&name, &token.location)?;
        self.advance_token();
        Ok(name)
    }

    fn parse_include_args(&mut self) -> Result<Vec<(String, VariableNode)>> {
        let mut args = Vec::new();
        let mut seen_keys = std::collections::HashSet::new();

        while matches!(self.current_token().kind, TokenKind::Whitespace(_)) {
            self.skip_whitespace();

            // Check if it's an identifier (not starting with /)
            let is_arg = match &self.current_token().kind {
                TokenKind::Ident(s) => !s.starts_with('/'),
                _ => false,
            };

            if !is_arg {
                break;
            }

            let key_token = self.consume_ident()?;
            let key = match key_token.kind {
                TokenKind::Ident(s) => s,
                _ => {
                    return Err(NatsuzoraError::ParseError {
                        message: format!(
                            "Internal error: expected identifier, got {:?}",
                            key_token.kind
                        ),
                        location: key_token.location,
                    });
                }
            };

            validator::validate_identifier(&key, &key_token.location)?;

            if seen_keys.contains(&key) {
                return Err(NatsuzoraError::ParseError {
                    message: format!("Duplicate include argument: {}", key),
                    location: key_token.location,
                });
            }
            seen_keys.insert(key.clone());

            self.skip_whitespace();
            self.consume(TokenKind::Equal)?;
            self.skip_whitespace();
            let value = self.parse_path()?;
            args.push((key, value));
        }

        Ok(args)
    }

    fn parse_variable_node(&mut self) -> Result<Node> {
        let path = self.parse_path()?;
        self.skip_whitespace();
        self.consume(TokenKind::Close)?;
        Ok(Node::Variable(path))
    }

    fn parse_path(&mut self) -> Result<VariableNode> {
        let first_token = self.current_token().clone();
        let mut segments = vec![self.parse_identifier_with_validation()?];

        while matches!(self.current_token().kind, TokenKind::Dot) {
            self.consume(TokenKind::Dot)?;
            segments.push(self.parse_identifier_with_validation()?);
        }

        Ok(VariableNode {
            path: segments,
            location: first_token.location,
        })
    }

    fn parse_identifier_with_validation(&mut self) -> Result<String> {
        let token = self.current_token().clone();

        // Check if it's a keyword (reserved word used as identifier)
        if self.is_keyword_token(&token) {
            self.advance_token();
            let word = self.keyword_to_string(&token.kind);
            return Err(NatsuzoraError::ReservedWordError {
                word,
                location: token.location,
            });
        }

        let token = self.consume_ident()?;
        let name = match &token.kind {
            TokenKind::Ident(s) => s.clone(),
            _ => {
                return Err(NatsuzoraError::ParseError {
                    message: format!("Internal error: expected identifier, got {:?}", token.kind),
                    location: token.location,
                });
            }
        };

        validator::validate_identifier(&name, &token.location)?;
        Ok(name)
    }

    fn is_keyword_token(&self, token: &Token) -> bool {
        matches!(
            token.kind,
            TokenKind::KwIf
                | TokenKind::KwUnless
                | TokenKind::KwElse
                | TokenKind::KwEach
                | TokenKind::KwAs
                | TokenKind::KwUnsecure
        )
    }

    fn keyword_to_string(&self, kind: &TokenKind) -> String {
        match kind {
            TokenKind::KwIf => "if".to_string(),
            TokenKind::KwUnless => "unless".to_string(),
            TokenKind::KwElse => "else".to_string(),
            TokenKind::KwEach => "each".to_string(),
            TokenKind::KwAs => "as".to_string(),
            TokenKind::KwUnsecure => "unsecure".to_string(),
            _ => "unknown".to_string(),
        }
    }

    fn block_close(&mut self, keyword: Option<TokenKind>) -> Result<bool> {
        if !matches!(self.current_token().kind, TokenKind::Open) {
            return Ok(false);
        }

        let saved_pos = self.pos;
        self.advance_token(); // open

        // If there's whitespace after {[, it cannot be a block close
        if matches!(self.current_token().kind, TokenKind::Whitespace(_)) {
            self.pos = saved_pos;
            return Ok(false);
        }

        let result = if matches!(self.current_token().kind, TokenKind::Slash) {
            if let Some(kw) = keyword {
                self.advance_token(); // slash
                self.skip_whitespace();
                self.current_token().kind == kw
            } else {
                true
            }
        } else {
            false
        };

        self.pos = saved_pos;
        Ok(result)
    }

    fn consume_block_close(&mut self, keyword: Option<TokenKind>) -> Result<()> {
        self.consume(TokenKind::Open)?;
        self.skip_whitespace();
        self.consume(TokenKind::Slash)?;
        self.skip_whitespace();
        if let Some(kw) = keyword {
            self.consume(kw)?;
        }
        self.skip_whitespace();
        self.consume(TokenKind::Close)?;
        Ok(())
    }

    fn current_token(&self) -> &Token {
        &self.tokens[self.pos]
    }

    fn is_eof(&self) -> bool {
        matches!(self.current_token().kind, TokenKind::Eof)
    }

    fn advance_token(&mut self) {
        if self.pos < self.tokens.len() - 1 {
            self.pos += 1;
        }
    }

    fn consume(&mut self, expected: TokenKind) -> Result<Token> {
        let token = self.current_token().clone();
        if std::mem::discriminant(&token.kind) != std::mem::discriminant(&expected) {
            return Err(NatsuzoraError::ParseError {
                message: format!("Expected {:?}, got {:?}", expected, token.kind),
                location: token.location,
            });
        }
        self.advance_token();
        Ok(token)
    }

    fn consume_text(&mut self) -> Result<Token> {
        let token = self.current_token().clone();
        if !matches!(token.kind, TokenKind::Text(_)) {
            return Err(NatsuzoraError::ParseError {
                message: format!("Expected text, got {:?}", token.kind),
                location: token.location,
            });
        }
        self.advance_token();
        Ok(token)
    }

    fn consume_ident(&mut self) -> Result<Token> {
        let token = self.current_token().clone();
        if !matches!(token.kind, TokenKind::Ident(_)) {
            return Err(NatsuzoraError::ParseError {
                message: format!("Expected identifier, got {:?}", token.kind),
                location: token.location,
            });
        }
        self.advance_token();
        Ok(token)
    }

    fn consume_required_whitespace(&mut self) -> Result<()> {
        if !matches!(self.current_token().kind, TokenKind::Whitespace(_)) {
            return Err(NatsuzoraError::ParseError {
                message: "Expected whitespace".to_string(),
                location: self.current_token().location,
            });
        }
        self.skip_whitespace();
        Ok(())
    }

    fn skip_whitespace(&mut self) {
        while matches!(self.current_token().kind, TokenKind::Whitespace(_)) {
            self.advance_token();
        }
    }

    fn unexpected_token<T>(&self, message: Option<&str>) -> Result<T> {
        let token = self.current_token();
        let msg = match message {
            Some(m) => format!("{}: {:?}", m, token.kind),
            None => format!("Unexpected token: {:?}", token.kind),
        };
        Err(NatsuzoraError::ParseError {
            message: msg,
            location: token.location,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;

    fn parse(source: &str) -> Result<Template> {
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize()?;
        let mut parser = Parser::new(tokens);
        parser.parse()
    }

    #[test]
    fn test_parse_text() {
        let template = parse("Hello, world!").unwrap();
        assert_eq!(template.nodes.len(), 1);
        if let Node::Text(node) = &template.nodes[0] {
            assert_eq!(node.content, "Hello, world!");
        } else {
            panic!("Expected Text node");
        }
    }

    #[test]
    fn test_parse_variable() {
        let template = parse("{[ name ]}").unwrap();
        assert_eq!(template.nodes.len(), 1);
        if let Node::Variable(node) = &template.nodes[0] {
            assert_eq!(node.path, vec!["name"]);
        } else {
            panic!("Expected Variable node");
        }
    }

    #[test]
    fn test_parse_path() {
        let template = parse("{[ user.profile.name ]}").unwrap();
        if let Node::Variable(node) = &template.nodes[0] {
            assert_eq!(node.path, vec!["user", "profile", "name"]);
        } else {
            panic!("Expected Variable node");
        }
    }

    #[test]
    fn test_parse_if_block() {
        let template = parse("{[#if visible]}Hello{[/if]}").unwrap();
        if let Node::IfBlock(node) = &template.nodes[0] {
            assert_eq!(node.condition.path, vec!["visible"]);
            assert_eq!(node.then_nodes.len(), 1);
            assert!(node.else_nodes.is_none());
        } else {
            panic!("Expected IfBlock node");
        }
    }

    #[test]
    fn test_parse_if_else_block() {
        let template = parse("{[#if visible]}Yes{[#else]}No{[/if]}").unwrap();
        if let Node::IfBlock(node) = &template.nodes[0] {
            assert!(node.else_nodes.is_some());
            assert_eq!(node.else_nodes.as_ref().unwrap().len(), 1);
        } else {
            panic!("Expected IfBlock node");
        }
    }

    #[test]
    fn test_parse_unless_block() {
        let template = parse("{[#unless hidden]}visible{[/unless]}").unwrap();
        if let Node::UnlessBlock(node) = &template.nodes[0] {
            assert_eq!(node.condition.path, vec!["hidden"]);
            assert_eq!(node.body_nodes.len(), 1);
        } else {
            panic!("Expected UnlessBlock node");
        }
    }

    #[test]
    fn test_parse_nested_unless_blocks() {
        let template = parse("{[#unless a]}{[#unless b]}inner{[/unless]}{[/unless]}").unwrap();
        if let Node::UnlessBlock(outer) = &template.nodes[0] {
            assert_eq!(outer.body_nodes.len(), 1);
            if let Node::UnlessBlock(inner) = &outer.body_nodes[0] {
                assert_eq!(inner.condition.path, vec!["b"]);
            } else {
                panic!("Expected inner UnlessBlock node");
            }
        } else {
            panic!("Expected outer UnlessBlock node");
        }
    }

    #[test]
    fn test_parse_each_block() {
        let template = parse("{[#each items as item]}{[ item ]}{[/each]}").unwrap();
        if let Node::EachBlock(node) = &template.nodes[0] {
            assert_eq!(node.collection.path, vec!["items"]);
            assert_eq!(node.item_name, "item");
            assert!(node.index_name.is_none());
        } else {
            panic!("Expected EachBlock node");
        }
    }

    #[test]
    fn test_parse_each_with_index() {
        let template = parse("{[#each items as item, idx]}{[ idx ]}{[/each]}").unwrap();
        if let Node::EachBlock(node) = &template.nodes[0] {
            assert_eq!(node.index_name, Some("idx".to_string()));
        } else {
            panic!("Expected EachBlock node");
        }
    }

    #[test]
    fn test_reserved_word_error() {
        let result = parse("{[ if ]}");
        assert!(result.is_err());
        if let Err(NatsuzoraError::ReservedWordError { word, .. }) = result {
            assert_eq!(word, "if");
        } else {
            panic!("Expected ReservedWordError");
        }
    }
}

File: typescript/src/parser.ts
/**
 * Natsuzora Parser
 *
 * Parses tokens into an AST.
 * Functional implementation without classes.
 */

import { ParseError, ReservedWordError } from "./errors.ts";
import { Token, TokenType, RESERVED_WORDS } from "./token.ts";
import {
  Node,
  Template,
  Path,
  IncludeArg,
  VariableModifier,
  createPath,
  textNode,
  variableNode,
  ifBlockNode,
  unlessBlockNode,
  eachBlockNode,
  unsecureOutputNode,
  includeNode,
  template,
} from "./ast.ts";
import { validateIncludeName } from "./validator.ts";

interface ParserState {
  tokens: Token[];
  pos: number;
}

function createState(tokens: Token[]): ParserState {
  // Filter out whitespace tokens for easier parsing
  return {
    tokens: tokens.filter((t) => t.type !== "whitespace"),
    pos: 0,
  };
}

function current(state: ParserState): Token {
  return state.tokens[state.pos] || { type: "eof", value: null, line: 0, column: 0 };
}

/**
 * Get the value of a token, throwing if null.
 * Use this after verifying the token type guarantees a value.
 */
function tokenValue(token: Token): string {
  if (token.value === null) {
    throw new ParseError(
      `Expected token value for ${token.type}`,
      token.line,
      token.column
    );
  }
  return token.value;
}

function peek(state: ParserState): Token | null {
  return state.tokens[state.pos + 1] || null;
}

function peekAt(state: ParserState, offset: number): Token | null {
  return state.tokens[state.pos + offset] || null;
}

function advance(state: ParserState): void {
  state.pos++;
}

function eof(state: ParserState): boolean {
  return state.pos >= state.tokens.length || current(state).type === "eof";
}

function expect(state: ParserState, type: TokenType): void {
  const token = current(state);
  if (token.type !== type) {
    throw new ParseError(`Expected ${type}, got ${token.type}`, token.line, token.column);
  }
  advance(state);
}

function expectKeyword(state: ParserState, type: TokenType, name: string): void {
  const token = current(state);
  if (token.type !== type) {
    throw new ParseError(`Expected '${name}', got ${token.type}`, token.line, token.column);
  }
  advance(state);
}

function validatePathSegment(name: string, line: number, column: number): void {
  if (name.startsWith("_")) {
    throw new ParseError(`Identifier '${name}' cannot start with underscore`, line, column);
  }
  if (RESERVED_WORDS.has(name)) {
    throw new ReservedWordError(name, line, column);
  }
}

function validateEachVariable(name: string, line: number, column: number): void {
  if (name.startsWith("_")) {
    throw new ParseError(`Identifier '${name}' cannot start with underscore`, line, column);
  }
  if (RESERVED_WORDS.has(name)) {
    throw new ReservedWordError(name, line, column);
  }
}

function parsePath(state: ParserState): Path {
  const segments: string[] = [];
  const firstToken = current(state);

  if (firstToken.type !== "ident") {
    throw new ParseError(`Expected identifier, got ${firstToken.type}`, firstToken.line, firstToken.column);
  }

  const firstName = tokenValue(firstToken);
  validatePathSegment(firstName, firstToken.line, firstToken.column);
  segments.push(firstName);
  advance(state);

  while (current(state).type === "dot") {
    advance(state);
    const segmentToken = current(state);
    if (segmentToken.type !== "ident") {
      throw new ParseError(`Expected identifier after '.', got ${segmentToken.type}`, segmentToken.line, segmentToken.column);
    }
    const segment = tokenValue(segmentToken);
    validatePathSegment(segment, segmentToken.line, segmentToken.column);
    segments.push(segment);
    advance(state);
  }

  return createPath(segments);
}

function parseModifier(state: ParserState): VariableModifier {
  if (current(state).type === "question") {
    advance(state);
    return "nullable";
  }
  if (current(state).type === "exclamation") {
    advance(state);
    return "required";
  }
  return null;
}

function parseVariable(state: ParserState): Node {
  const path = parsePath(state);
  const modifier = parseModifier(state);
  expect(state, "close");
  return variableNode(path, modifier);
}

function parseUnsecureOutput(state: ParserState): Node {
  expect(state, "bang_unsecure");
  const path = parsePath(state);
  expect(state, "close");
  return unsecureOutputNode(path);
}

function parseInclude(state: ParserState): Node {
  expect(state, "bang_include");

  const nameToken = current(state);
  const nameValue = nameToken.type === "ident" ? tokenValue(nameToken) : null;
  if (nameToken.type !== "ident" || !nameValue?.startsWith("/")) {
    throw new ParseError(`Expected include name starting with '/', got ${nameValue}`, nameToken.line, nameToken.column);
  }

  const name = nameValue;
  validateIncludeName(name, nameToken.line, nameToken.column);
  advance(state);

  const args: IncludeArg[] = [];

  while (current(state).type === "ident" && !tokenValue(current(state)).startsWith("/")) {
    const keyToken = current(state);
    const key = tokenValue(keyToken);

    if (key.startsWith("_")) {
      throw new ParseError(`Identifier '${key}' cannot start with underscore`, keyToken.line, keyToken.column);
    }

    if (args.some((a) => a.key === key)) {
      throw new ParseError(`Duplicate include argument: ${key}`, keyToken.line, keyToken.column);
    }

    advance(state);
    expect(state, "equal");
    const value = parsePath(state);
    args.push({ key, value });
  }

  expect(state, "close");
  return includeNode(name, args);
}

function parseIfBlock(state: ParserState): Node {
  expect(state, "kw_if");
  const condition = parsePath(state);
  expect(state, "close");

  const thenBranch = parseNodes(state);

  let elseBranch: Node[] | null = null;

  if (current(state).type === "open") {
    const next = peek(state);
    if (next?.type === "hash") {
      const afterHash = peekAt(state, 2);
      if (afterHash?.type === "kw_else") {
        expect(state, "open");
        expect(state, "hash");
        expect(state, "kw_else");
        expect(state, "close");
        elseBranch = parseNodes(state);
      }
    }
  }

  expect(state, "open");
  expect(state, "slash");
  expectKeyword(state, "kw_if", "if");
  expect(state, "close");

  return ifBlockNode(condition, thenBranch, elseBranch);
}

function parseUnlessBlock(state: ParserState): Node {
  expect(state, "kw_unless");
  const condition = parsePath(state);
  expect(state, "close");

  const body = parseNodes(state);

  expect(state, "open");
  expect(state, "slash");
  expectKeyword(state, "kw_unless", "unless");
  expect(state, "close");

  return unlessBlockNode(condition, body);
}

function parseEachBlock(state: ParserState): Node {
  expect(state, "kw_each");
  const collection = parsePath(state);
  expect(state, "kw_as");

  const itemToken = current(state);
  if (itemToken.type !== "ident") {
    throw new ParseError(`Expected identifier for each item, got ${itemToken.type}`, itemToken.line, itemToken.column);
  }
  const itemName = tokenValue(itemToken);
  validateEachVariable(itemName, itemToken.line, itemToken.column);
  advance(state);

  expect(state, "close");

  const body = parseNodes(state);

  expect(state, "open");
  expect(state, "slash");
  expectKeyword(state, "kw_each", "each");
  expect(state, "close");

  return eachBlockNode(collection, itemName, body);
}

function parseBlockOpen(state: ParserState): Node {
  expect(state, "hash");
  const token = current(state);

  switch (token.type) {
    case "kw_if":
      return parseIfBlock(state);
    case "kw_unless":
      return parseUnlessBlock(state);
    case "kw_each":
      return parseEachBlock(state);
    default:
      throw new ParseError(`Unknown block type: ${token.type}`, token.line, token.column);
  }
}

function isKeywordToken(type: string): boolean {
  return type.startsWith("kw_");
}

function keywordToName(type: string): string {
  // "kw_if" -> "if", "kw_each" -> "each", etc.
  return type.slice(3);
}

function parseTag(state: ParserState): Node {
  expect(state, "open");
  const token = current(state);

  switch (token.type) {
    case "hash":
      return parseBlockOpen(state);
    case "bang_unsecure":
      return parseUnsecureOutput(state);
    case "bang_include":
      return parseInclude(state);
    case "ident":
      return parseVariable(state);
    default:
      // Check if it's a keyword being used as identifier
      if (isKeywordToken(token.type)) {
        throw new ReservedWordError(keywordToName(token.type), token.line, token.column);
      }
      throw new ParseError(`Unexpected token in tag: ${token.type}`, token.line, token.column);
  }
}

function parseNode(state: ParserState): Node | null {
  const token = current(state);

  switch (token.type) {
    case "text":
      advance(state);
      return textNode(tokenValue(token));
    case "open":
      return parseTag(state);
    case "eof":
      return null;
    default:
      throw new ParseError(`Unexpected token: ${token.type}`, token.line, token.column);
  }
}

function parseNodes(state: ParserState): Node[] {
  const nodes: Node[] = [];

  while (!eof(state)) {
    const token = current(state);

    // Check for block close or else
    if (token.type === "open") {
      const next = peek(state);
      if (next?.type === "slash" || next?.type === "kw_else") {
        break;
      }
      if (next?.type === "hash") {
        const afterHash = peekAt(state, 2);
        if (afterHash?.type === "kw_else") {
          break;
        }
      }
    }

    const node = parseNode(state);
    if (node) {
      nodes.push(node);
    }
  }

  return nodes;
}

/**
 * Parse tokens into an AST Template.
 */
export function parse(tokens: Token[]): Template {
  const state = createState(tokens);
  const nodes = parseNodes(state);
  return template(nodes);
}
